### ISSUE TITLE
torch.autograd profiler changes the model's outputs

### ISSUE THREAD

**Issue by sdascoli** [pytorch/pytorch#59941]
## 🐛 Bug

I recently encountered trouble as the accuracy of models loaded from timm package were different from one evaluration script to the other.
I ended up realizing that the difference was coming from the torch.autograd profiler which I was calling before the evaluation.
How come calling the profiler changes the outputs of the model ?

## To Reproduce

Here is a minimal example to reproduce the bug:

```
@torch.no_grad()
def profile_model(model):
    y = torch.randn(1,3,224,224).cuda()
    with torch.autograd.profiler.profile(use_cuda=True) as prof:
        model(y)
    return prof

x = torch.randn(1,3,224,224).cuda()

model = resnet18(pretrained=True).cuda()
profile_model(model)
model.eval()
print(model(x).norm())

model = resnet18(pretrained=True).cuda()
model.eval()
print(model(x).norm())
```

## Expected behavior

Since we use the same model, the output should be the same, but they are not.

## Environment

```
PyTorch version: 1.8.0
Is debug build: False
CUDA used to build PyTorch: 10.2
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.1 LTS (x86_64)
GCC version: (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0
Clang version: Could not collect
CMake version: version 3.16.3
Libc version: glibc-2.10

Python version: 3.7 (64-bit runtime)
Python platform: Linux-5.4.0-52-generic-x86_64-with-debian-bullseye-sid
Is CUDA available: True
CUDA runtime version: 10.1.105
GPU models and configuration:
GPU 0: Quadro GP100
GPU 1: Quadro GP100

Nvidia driver version: 450.80.02
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.19.2
[pip3] torch==1.8.0
[pip3] torch-utils==0.1.2
[pip3] torchaudio==0.8.0a0+f945023
[pip3] torchvision==0.9.0.dev20210123
```

cc @ilia-cher @gdankel @ngimel @nbcsm @guotuofeng @guyang3532 @gaoteng-git

**Comment by soulitzer** [comment]
If you put your model in eval mode before running forward under the profiler, the two prints return the same result. So the issue is probably just that running statistics are being updated. Feel free to reopen/file a new issue if something else comes up.
